compositor_pipeline/src/pipeline.rs

```diff
@@ -321,129 +324,115 @@ impl Pipeline {
-    fn clone(&self) -> Self {
-        Self(self.0.clone())
-    }
-}
-
-impl<T> OutputRegistry<T> {
-    fn new() -> Self {
-        Self(Arc::new(Mutex::new(HashMap::new())))
-    }
-
-    fn contains_key(&self, key: &OutputId) -> bool {
-        self.0.lock().unwrap().contains_key(key)
-    }
-
-    fn insert(&self, key: OutputId, value: Arc<T>) -> Option<Arc<T>> {
-        self.0.lock().unwrap().insert(key, value)
-    }
-
-    fn remove(&self, key: &OutputId) -> Option<Arc<T>> {
-        self.0.lock().unwrap().remove(key)
-    }
-
-    fn lock(&self) -> MutexGuard<HashMap<OutputId, Arc<T>>> {
-        self.0.lock().unwrap()
-    }
-}
+fn run_renderer_thread(
+    pipeline: Arc<Mutex<Pipeline>>,
+    frames_receiver: Receiver<QueueVideoOutput>,
+) {
+    let renderer = pipeline.lock().unwrap().renderer.clone();
+    for mut input_frames in frames_receiver.iter() {
+        for (input_id, event) in input_frames.frames.iter_mut() {
+            if let PipelineEvent::EOS = event {
+                let mut guard = pipeline.lock().unwrap();
+                let eos_received = guard
+                    .inputs
+                    .get_mut(input_id)
+                    .and_then(|input| input.video_eos_received.as_mut());
+                if let Some(eos_received) = eos_received {
+                    *eos_received = true;
+                }
+            }
+        }
+        let output = renderer.render(input_frames.into());
+        let Ok(output_frames) = output else {
+            error!(
+                "Error while rendering: {}",
+                ErrorStack::new(&output.unwrap_err()).into_string()
+            );
+            continue;
+        };
 
-pub struct OutputIterator<'a> {
-    inner_iter: hash_map::Iter<'a, OutputId, Arc<PipelineOutput>>,
-}
+        for (id, frame) in output_frames.frames {
+            let frame_sender = pipeline
+                .lock()
+                .unwrap()
+                .outputs
+                .get(&id)
+                .and_then(|output| output.encoder.frame_sender())
+                .cloned();
+            let Some(frame_sender) = frame_sender else {
+                error!("no output with id {}", &id);
+                continue;
+            };
 
-impl<'a> OutputIterator<'a> {
-    fn new(iter: hash_map::Iter<'a, OutputId, Arc<PipelineOutput>>) -> Self {
-        Self { inner_iter: iter }
+            if frame_sender.send(PipelineEvent::Data(frame)).is_err() {
+                warn!("Failed to send output frames. Channel closed.")
+            }
+        }
     }
 }
 
-impl<'a> Iterator for OutputIterator<'a> {
-    type Item = (&'a OutputId, &'a PipelineOutput);
+fn run_audio_mixer_thread(
+    pipeline: Arc<Mutex<Pipeline>>,
+    audio_receiver: Receiver<QueueAudioOutput>,
+) {
+    let audio_mixer = pipeline.lock().unwrap().audio_mixer.clone();
+    for mut samples in audio_receiver.iter() {
+        for (input_id, event) in samples.samples.iter_mut() {
```

Comment:
This is only marking input as EOS, but we need to add logic that check the inputs and decides whether to send EOS on output too. I'll add that in the follow-up.